File: 2019/Day 2/Task 2-2.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>

#define CORRECT_RESULT 19690720

using namespace std;

int execute ( vector < int > data , int input_a, int input_b )
{
	data[1] = input_a;
	data[2] = input_b;
	int pos = 0;
	while( data[pos] != 99 )
	{
		switch( data[pos] ) {
			case 1 : data[data[pos + 3]] = data[data[pos + 1]] + data[data[pos + 2]];
					 break;
			case 2 : data[data[pos + 3]] = data[data[pos + 1]] * data[data[pos + 2]];
					 break;
			default : cout << "There has been a wrong input";
			}
		pos += 4;
	}
	return data[0];
}

vector < int > open()
{
	ifstream fin;
	vector < int > data;
	string x;
	fin.open("Task 2-1.txt");
	if( fin.is_open() )
	{
		while( getline (fin, x, ','))
		{
			data.push_back( stoi(x) );
		}
		fin.close();
	}
	else
		cout << "Can't open the file" << endl;
	return data;
}

int main ()
{
	vector < int > data;
	int noun = 0, verb = 0, result;
	
	data = open();
	result = execute( data, noun, verb );
	data.clear();
	
	while(result != CORRECT_RESULT){
		noun++;
		if( noun == 100 )
		{
			noun = 0;
			verb++;
		}
		data = open();
		result = execute( data, noun, verb );
		data.clear();
		
	}
	
	
	cout << 100 * noun + verb << endl;
}

File: 2019/Day 2/Task 2-2.py
def load_code():
    with open("Task 2-1.txt") as src:
        return list(map(int, src.read().split(",")))


def execute(program, noun, verb):
    pos = 0
    program[1] = noun
    program[2] = verb
    while pos < len(program):
        opcode = program[pos]
        if opcode == 1 or opcode == 2:
            op1 = program[pos + 1]
            op2 = program[pos + 2]
            dest = program[pos + 3]
            program[dest] = program[op1] + program[op2] if opcode == 1 else program[op1] * program[op2]
            pos += 4
        elif opcode == 99:
            break
        else:
            print("There has been a wrong input")
            break
    return program[0]


noun = 0
verb = -1
result = 0
while result != 19690720:
    verb += 1
    if verb == 100:
        verb = 0
        noun += 1
    intcode = load_code()
    result = execute(intcode, noun, verb)
print(100 * noun + verb)
